lib/matplotlib/figure.py

```diff
@@ -576,12 +576,12 @@ def add_axes(self, *args, **kwargs):
             if not np.isfinite(rect).all():
                 raise ValueError('all entries in rect must be finite '
                                  'not {}'.format(rect))
-            projection_class, kwargs = self._process_projection_requirements(
```

Comment:
Why the change in identifier here?

Reply:
I think that this should be changed back to `kwargs`.

Reply:
Well we are going to strip some kwargs, so it makes sense to return something different here I think?

Reply:
The stripping has already been done above, in this function. This call to `_process_projection_requirements` _normalizes_ the remaining keyword arguments. You probably want to save the _normalized_ keyword arguments. Changing this identifier actually changes the behavior of the code in a very minor but probably undesirable way.

Reply:
The issue is that we need to do the check in pyplot on the stashed values to determine if we need to enter this function or not so I think we need to cache the un-processed kwargs (which is what we were doing for 3.3 and prior).

There is the issue that we pop `'rect'` out of the kwargs, however `plt.subplot` forces `args` to be non-empty and `add_axes` will raise if you pass both 'rect' and non-empty args so we can never drive to this case.

Reply:
By "this function," do you mean `_process_projection_requirements` or `_add_axes_internal`?

Reply:
I mean `Figure.add_axes`.  The only place we have this state is in `plt.subplot` where the logic is:

 - use args to compute _positional_ duplication
 - if one is found, check the kwargs
   - if user passed kwargs match the stashed kwargs, use them
   - if they do not match, make a new one via `Figure.add_axes` and stash the new user passed kwargs.

Reply:
Oh, I understand now.

Note, the logic that maps `polar=True` and `projection='polar'` to the same key lives in `_process_projection_requirements`. So in the past, these two statements would have returned the same axes:

```python
ax1 = plt.subplot(111, projection='polar')
ax2 = plt.subplot(111, polar=True)
assert ax1 is ax2  # passes for Matplotlib 3.3.4, but will not pass with this patch
```

I think that this normalization logic has to be lifted out into `plt.subplot` if we want this to work still in this one special case.

Reply:
drat :(

Reply:
Pushed new changes so that we now track the projection class + post processed kwargs.  Added @lpsinger 's code as a tost case (thank you :) ).

Reply:
Looks good now!

Reply:
The point of keeping `plt.subplot` ability to grab the old axes was to give pyplot users a way to get the handle that they were otherwise lacking.   I am not convinced that adding extra logic to handle the above case was a good idea in order to support retrieving the axes.

Reply:
The new version calls `_process_projection_requirements` twice: once in `plt.subplot`, and then again in `Figure.add_subplot`, which is called by the former. As long as we are OK with that very minor inefficiency, this looks good.

Reply:
My concern is that the previous version was simpler (at least to me), and I don't know know that we need to support recovering the axes using two different spellings for polar.  Overall our code base has lots of clever things in it that support a bunch of weird edge cases like this that makes it very difficult to follow what the code is doing.  If we can avoid that as much as is reasonable, it would be great.

Reply:
I don't see anyway around the inefficiency (short of some pretty major refactors).

I agree it is annoying we have 2 ways of spelling polar, but it used to work and "it is annoying to normalize" is not a compelling reason to break back-compatibility.

The final version of this is going to be a fair bit more complicated, but at least all of that complication is in `pyplot` rather than spread through the `Figure` codebase.